sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,672 @@
+  not listed in the attribute.  However, this check only results in a warning,
+  so the runtime is still responsible for throwing the exception if any of the
+  functions called by the kernel uses an optional feature that the device does
+  not support.
+
+* For a kernel that is decorated with the `[[sycl::reqd_work_group_size(W)]]`
+  or `[[sycl::reqd_sub_group_size(S)]]` attribute, the exception must be thrown
+  if the device does not support the work group size `W` or the sub-group size
+  `S`.
+
+Note that the exception must be thrown synchronously, not delayed and thrown on
+the queue's asynchronous handler.
+
+
+### No runtime exception for speculative compilation
+
+It is currently common for the runtime to speculatively compile some kernels.
+For example, DPC++ may bundle all kernels from the same translation unit
+together into a single device image.  When the application submits one kernel
+*K* to a device *D*, the runtime actually compiles all kernels in *K*'s device
+image for device *D*.  Let's assume in this example that the kernel *K* uses
+only features that are supported by *D*.  It would be illegal for the runtime
+to throw an exception in such a case just because some other kernel in the same
+device image uses a feature that is not supported by device *D*.
+
+
+## Design
+
+### Changes to DPC++ headers
+
+With the exception of the `[[sycl::reqd_work_group_size()]]` and
+`[[sycl::reqd_sub_group_size()]]` attributes, all kernel optional features
+are associated with some device aspect.  For example, the `sycl::half` type
+is an optional feature which is only supported on devices that have the
+`aspect::fp16` aspect.  We can therefore use device aspects as a way to
+describe the set of optional features that a kernel uses (with the exception
+of the required work-group or sub-group size).
+
+As we will see later, it is helpful to decorate all APIs in DPC++ headers that
+correspond to optional kernel features with a C++ attribute that identifies the
+associated aspect.  We cannot use the `[[sycl::requires()]]` attribute for this
+purpose, though, because that attribute is allowed only for functions.
+Instead, we invent a new internal attribute `[[sycl_detail::uses_aspects()]]`
+that can be used to decorate both functions and types.  This attribute is not
+documented for user code; instead it is an internal implementation detail of
+DPC++.
+
+Like all use of C++ attributes in the DPC++ headers, the headers should spell
+the attribute using initial and trailing double underscores
+(`[[__sycl_detail__::__uses_aspects__()]]`).  We show that form in the code
+samples below, but this design document uses the form without the underbars
+elsewhere.  Both forms refer to the same attribute.
+
+To illustrate, the type `sycl::half` is an optional feature whose associated
+aspect is `aspect::fp16`.  We therefore decorate the declaration like this:
+
+```
+using half [[__sycl_detail__::__uses_aspects__(has(aspect::fp16))]] =
```

Comment:
I did this mostly to reuse the syntax that we need already for the `[[sycl::requires()]]` attribute.  I figured it would be easier for the front-end to implement if we use the same parameter syntax for both attributes.  If that is not the case, then I agree that we could simplify the parameter syntax for `[[sycl_detail::uses_aspects()]]`.  It would probably be wise, though, to at least allow a list of aspects, because I could foresee the possibility that some optional features might require a device with more than one aspect.

Since you will probably ask ... the reason the `[[sycl::requires()]]` attribute has the `has()` clause is because we think we may want to extend this in the future to allow boolean expressions of required aspects.  For example, we could imagine a syntax like `[[sycl::requires(has(aspect::foo) || has(aspect::bar))]]`.  We think this syntax might be useful once we add the [device-if](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc) feature.  However, I do not foresee a need for such complicated expressions for `[[sycl_detail::uses_aspects()]]`.

Reply:
We talked about the syntax for this attribute in our meeting yesterday, and we are considering simplifying the parameter list.  @AaronBallman: can we assume that the simplified syntax I describe below would be easier for the FE to deal with?

Some brief motivation ... I don't like the fact that the common use case of `sycl::requires()` needs the `has()` clause.  Even if we do allow disjunctions in the future, they will be extremely uncommon.  Therefore, I'd like the common case to have simpler syntax like `sycl::requires(aspect1, aspects2)`.  If we are changing this part of the spec anyways, we'd also like to rename the attribute to avoid confusion with the C++20 "requires" keyword.

Therefore, we are thinking of changing the `[[sycl::requires()]]` attribute to be like: `[[sycl::device_has(aspect, aspect, ...)]]`.  For example:

```
[[sycl::device_has(aspect::fp64, aspect::atomic64)]]
```

If we do this, then we would also change the internal attribute `[[sycl::uses_aspects()]]` to have the same parameter syntax.

Reply:
I'll leave it to @AaronBallman to comment on the new name but yes removing the `has`  clause simplifies FE implementation.

Reply:
>  @AaronBallman: can we assume that the simplified syntax I describe below would be easier for the FE to deal with?

Yes, this would simplify things greatly! But if you do have disjunctions, what would they look like? (Just trying to make sure we don't paint ourselves into a corner with the design.)

> If we are changing this part of the spec anyways, we'd also like to rename the attribute to avoid confusion with the C++20 "requires" keyword.

Aside: this one turns out to be not super problematic because C++ requires implementations to treat keywords within an attribute list as though they were identifiers (https://eel.is/c++draft/dcl.attr#grammar-4.sentence-5).

Reply:
> Aside: this one turns out to be not super problematic because C++ requires implementations to treat keywords within an attribute list as though they were identifiers

I think it still makes sense to rename the attribute for the sake of humans who will read the code.  Using a name that matches an unrelated keyword seems like it might cause confusion.

> But if you do have disjunctions, what would they look like? (Just trying to make sure we don't paint ourselves into a corner with the design.)

Here is the rationale I plan to pitch to the committee:

> Our original intent was to allow the `[[sycl::requires()]]` attribute to express both conjunctions and disjunctions of aspects in some future version of the SYCL spec.  For example, we thought we might want to allow syntax like this in the future:
>
> ```
> [[sycl::requires(has(aspect::fp64, aspect::atomic64) || has(aspect::fp16))]]
> ```
>
> We still think this might be desirable in the future, but we think disjunctions like this will be very uncommon.  We don't want to incur the complexity of the `has()` clause even in the common case, so we omit it from the current syntax (which only supports conjunctions of aspects).
>
> If we do decide to support disjunction in the future, we can create a new attribute that supports the expanded syntax. 
 For example:
>
> ```
> [[sycl::device_has_expression(has(aspect::fp64, aspect::atomic64) || has(aspect::fp16))]]
> ```
>
> (We can bikeshed the name `device_has_expression` later if we decide to add this.)

Reply:
> I think it still makes sense to rename the attribute for the sake of humans who will read the code. Using a name that matches an unrelated keyword seems like it might cause confusion.

No complaints from me on that! I just didn't want you to think you had to rename it because the keyword was an issue.

> If we do decide to support disjunction in the future, we can create a new attribute that supports the expanded syntax.
> For example:
> `[[sycl::device_has_expression(has(aspect::fp64, aspect::atomic64) || has(aspect::fp16))]]`
> (We can bikeshed the name device_has_expression later if we decide to add this.)

This would run into the same FE implementation issues I was worried about. To explain in a bit more detail: Clang (and I believe other implementations as well) implement parsing attributes in a generic way: we see `[[` in a place where we expect an attribute specifier to appear, and we parse the attribute name (like `sycl::requires`) up to the argument list. Once we know the attribute, we can decide how to parse its argument list. Currently, the *vast* majority of attribute argument lists are "simple" in that they can all be parsed generically based off information we have in a tablegen file. (The tablegen file says "you have these attributes, they expect this number of arguments of this type, etc" and the arguments are typically "a string" or "an expression" or "one of these identifiers as though it was an enumeration", etc.) A later stage takes these parsed attributes and turns them into AST nodes while checking that what was parsed is valid based on the same information in the tablegen file. These `has` clauses are a totally different thing that we basically have to write custom parsing logic to support. That custom parsing logic isn't actually the bad part of this -- we do custom parsing already in other circumstances. The bad part is taking the parsed attributes and turning them into AST nodes -- that generic "parsed attribute" interface knows about simple things like expressions, identifiers, etc. It has no notion of has clauses or predicate expressions between clauses like `||` and so we'd have to get inventive there. What's more, because the predicate expressions aren't on actual C or C++ syntax, we have to make our own expression evaluator to evaluate the logic. (e.g., `[[foo(false || 12 + i == 17)]]` is easy to evaluate assuming `i` is an integer constant expression because we can just pass the `Expr *` held by the parsed attribute to the constant expression evaluator. But `[[foo(has(false) || has(12 + i == 17))]]` is hard when `has` is part of the attribute syntax because there's no `Expr *` to evaluate.)

Reply:
It might be better to debate this is detain later if / when we decide to extend the parameters to allow disjunctions.  It's possible that the extension won't even use C++ attributes since we are considering an alternate syntax based on "properties".

Still, I don't want to paint ourselves into a corner in case we decide to stick with a C++ attribute syntax.  My main concern at this point is to make sure we don't make things worse in the future by changing from `[[sycl::requires(has(asp, asp, ...))]]` to `[[sycl::device_has(asp, asp, ...)]]`.  At this point, I don't see how this change makes things harder for us in the future.

At the risk of going into too much depth now, I did want to address one concern.  Even if we do add disjunctions, we wouldn't want the CFE to do any sort of constant expression evaluation.  We would just want some straightforward transcription of the parameters into LLVM metadata, and later phases would be responsible for using the information.  Hopefully, that would avoid the need for extending the structure of AST nodes.

Reply:
> At this point, I don't see how this change makes things harder for us in the future.

It doesn't make it harder for us in the future; it's the same level of hard as today.

> Hopefully, that would avoid the need for extending the structure of AST nodes.

It wouldn't avoid that need, but we can cross that bridge when we come to it.

Reply:
I submitted [PR171](https://github.com/KhronosGroup/SYCL-Docs/pull/171) against the SYCL 2020 spec to change the name of the `[[sycl::requires()]]` aspect to `[[sycl::device_has()]]` and to remove the `has()` clause.

In anticipation of that being adopted, I removed the `has()` clause from `[[sycl::uses_aspects()]]` in 1e927c107a7ac5c486a5620fba4ce80912701785.  I'll rename the `[[sycl::requires()]]` references to `[[sycl::device_has()]]` when/if [PR171](https://github.com/KhronosGroup/SYCL-Docs/pull/171) is adopted.